interface.go

```diff
@@ -0,0 +1,74 @@
+// Package provider represents a reference implementation of an index provider.
+package provider
+
+import (
+	"context"
+
+	"github.com/filecoin-project/storetheindex/api/v0/ingest/schema"
+	"github.com/ipfs/go-cid"
+	mh "github.com/multiformats/go-multihash"
+)
+
+// LookupKey represents the key that uniquely identifies a list of multihashes looked up via Callback.
+// See: Interface.NotifyPut, Interface.NotifyRemove, ProviderCallback.
+type LookupKey []byte
+
+// Interface represent an index provider that manages the advertisement of multihashes to indexer nodes.
+// See:
+// - https://pkg.go.dev/github.com/filecoin-project/go-indexer-core
+// - https://pkg.go.dev/github.com/filecoin-project/storetheindex
+type Interface interface {
+
+	// PublishLocal appends adv to the locally stored advertisement chain and returns the corresponding CID to it.
+	// This function does not publish the changes to the advertisement chain onto gossip pubsub channel.
+	// Use Publish instead if indexer nodes must be made aware of the appended advertisement.
+	//
+	// See: Publish.
+	PublishLocal(ctx context.Context, adv schema.Advertisement) (cid.Cid, error)
+
+	// Publish appends adv to the locally stored advertisement chain, and publishes the new advertisement onto gossip pubsub.
+	// The CID returned represents the ID of the advertisement appended to the chain.
+	Publish(ctx context.Context, adv schema.Advertisement) (cid.Cid, error)
+
+	// RegisterCallback registers the callback used by the provider to look up a list of multihashes by LookupKey.
+	// Only a single callback is supported; repeated calls to this function will replace the previous callback.
+	RegisterCallback(cb Callback)
+
+	// NotifyPut sginals to the provider that the list of multihashes looked up by the given key are available.
+	// The given key is then used to look up the list of multihashes via Callback.
+	// An advertisement is then generated, appended to the chain of advertisements and published onto the gossip pubsub channel.
+	// Therefore, a Callback must be registered prior to using this function.
+	//
+	// The metadata is serialized data that provides hits about how to retrieve data and is entirely protocol dependant.
+	// The metadata may be nil and is optional.
+	// If specified, the metadata value must start with a uvarint that correspons to the indexer.Value protocol ID.
+	//
+	// This function returns the ID of the advertisement published.
+	//
+	// See: https://pkg.go.dev/github.com/filecoin-project/go-indexer-core
+	// TODO: update pkg link above once encodeMetadata is exported and docs are updated.
```

Comment:
@gammazero  Ieft a TODO here for myself to update the link once encodeMetadata is exported. Please shout if you can think of a better URL to put in there just now. Cheers